CI: Fix lint action and fail if warnings are present.

mirenradia opened the pull request:
This PR does the following:
* Fixes #56 by using `v2` of the cpp-linter-action.
* Updates the lint workflow to fail if there are warnings.
* Updates the lint workflow to run on the `pull_request` trigger.
* Removes the style workflow and uses the clang-format checking provided by cpp-linter-action in the lint workflow.
* Updates the lint workflow to post a comment on PRs with the warnings outputted by clang-tidy and the incorrectly formatted files as determined by clang-format (an example can probably be found below).
* Updates the LLVM version used (for clang-tidy and clang-format) to 17 (the pre-commit config has also been updated).
* Updates the lint workflow to use a compilation database created by [compiledb](https://github.com/nickdiego/compiledb) rather than parsing the output of e.g. `make print-CXXFLAGS` and adding these as extra flags. Note that this means that files which are not in the compilation database (i.e. are not currently in a test or example) will not have any flags associated with them and there will be warnings.

Note that our `.impl.hpp` template implementation files will always fail linting as they cannot be compiled on their own. I have chosen not to pass them to the `ignore` argument as we still want to see other other valid warnings from these files and we also want them to be checked for formatting.

This resolves #12.

mirenradia committed:
CI: Update lint action

Make action fail if clang-tidy prints warnings.

mirenradia committed:
CI: Make lint action ignore AMReX

mirenradia committed:
CI: Ignore External and add tests includes for lint action

mirenradia committed:
CI: Enable formatting in lint action

Also enable printing output in the workflows summary.

mirenradia committed:
CI: Update lint action

Add some more comments and clean up naming. Disable cleaning for
subaction.

mirenradia committed:
CI: Remove style action

Clang-format is now handled by the lint action.

mirenradia committed:
Lint: Fix clang-tidy warnings

mirenradia committed:
Lint: Ignore unconditional preprocessor if warnings

mirenradia committed:
Lint/Format: Update clang-tidy/format to v17

mirenradia committed:
Lint: Use clang-format style when clang-tidy fixing

mirenradia committed:
CI: Switch to compilation database for lint action

mirenradia committed:
AMRInterpolator: Add missing implementation file include guards

mirenradia committed:
Make: Update clang-tidy targets to lint header files

Explicitly exclude implementation and include header files as these can't be
compiled on their own.

mirenradia commented:
As described above, it is expected that the lint workflow will fail if `.impl.hpp` files are modified so it is fine to approve this with that check failing, @julianakwan.

julianakwan commented:
Hi Miren - I was looking through the files that had been changed and I noticed that lint changed some virtual void functions in GRAMR and GRAMRLevel to just void, e.g. post_restart(). Is this because there was nothing further down the inheritance chain that required them to be virtual? I supposed if there was a need for them to be virtual later on, we could change it by hand.

mirenradia commented:
> Hi Miren - I was looking through the files that had been changed and I noticed that lint changed some virtual void functions in GRAMR and GRAMRLevel to just void, e.g. post_restart(). Is this because there was nothing further down the inheritance chain that required them to be virtual? I supposed if there was a need for them to be virtual later on, we could change it by hand.

Actually, it's still possible to override the virtual function from the base classes (`Amr` and `AmrLevel` respectively) since we're using `override` rather than `final`. The C++ Core Guidelines recommends using one of `virtual`, `override` or `final` for virtual functions (see [here](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-override)) hence why clang-tidy removed the redundant `virtual`.

mirenradia committed:
Make: Fix single file clang-tidy target

This should now work when the file is not in the current directory and
also allows clang-tidy to check header files.
